buzzmobile/tests/test_utils/rostest_utils.py

```diff
@@ -1,38 +1,74 @@
 """A collection of utilities to make testing with ros less painful.
 """
 
+import os
+import random
 import subprocess
+import unittest
 
-def with_roscore(obj):
-    """Decorator to run all tests in a testcase with their own roscore.
 
-    This wraps the setUp and tearDown methods to start by first spinning up a
-    roscore process, and tears it down at the very end. This adds a small time
-    penalty, but its worth it.
+def rand_port():
+    """Picks a random port number.
 
-    Its worth deciding whether to make this work on a per-method, per object,
-    or both basis.
+    This is potentially unsafe, but shouldn't generally be a problem.
     """
-    old_setup = obj.setUp
-    old_teardown = obj.tearDown
-
-    def new_setup(self):
-        """Wrapper around the user-defined setUp method that runs roscore.
-        """
-        self.roscore = subprocess.Popen(['roscore'])
-        old_setup(self)
-
-    def new_teardown(self):
-        """Wrapper around the user-defined tearDown method that ends roscore.
-        """
-        old_teardown(self)
-        self.roscore.kill()
-        subprocess.call(['killall', '-9', 'rosmaster'])
-        self.roscore.wait()
-
-    obj.setUp = new_setup
-    obj.tearDown = new_teardown
-    return obj
+    return random.randint(10311, 12311)
+
+
+class RosTestMeta(type):
+    def __new__(cls, name, bases, dct):
+
+        # It will break unless we throw in fake setup and teardown methods if
+        # the real ones don't exist yet.
+
+        def fake_settear(self):
+            _ = self
+
+        try:
+            old_setup = dct['setUp']
+        except KeyError:
+            old_setup = fake_settear
+        try:
+            old_teardown = dct['tearDown']
+        except KeyError:
+            old_teardown = fake_settear
+
+        def new_setup(self):
+            """Wrapper around the user-defined setUp method that runs roscore.
+            """
+            self.port = rand_port()
```

Comment:
yes, hence the todo to retry until success, which should just be (in pseudocode):

    port = randport
    while port is in use:
        port = randport